### PR TITLE
Added full command support

### PR DESCRIPTION
**Pull request by softwarecrash**
- Added all commands from the datasheet V 1.2
- Added undocumented Charge and Discharge FET Controll
- make compatible with Arduino / ESP Environment 

Known bugs
- peak Current false values (not sure it comes from the bms or its wrong reading/calculating) - Working on it

i hope you can rewrite the readme file, my english its not the best to wrote this clearly

### CODE REVIEW

File: daly-bms-uart.cpp [hunk 1]
```diff
@@ -41,27 +38,45 @@ bool Daly_BMS_UART::Init()
     return true;
 }
 
-bool Daly_BMS_UART::update()
+bool Daly_BMS_UART::updateAllValues()
 {
-
+    // There are better ways to write this, but this is fine...
```
Comment (softwarecrash) [inline daly-bms-uart.cpp:43]: I don't Know a netter was, please Tell me to so it right

File: daly-bms-uart.h [hunk 2]
```diff
@@ -143,9 +161,9 @@ class Daly_BMS_UART
     bool Init();
 
     /**
-     * send data to the bms and update the values
+     * @brief Update all the values in the get struct
      */
-    bool update();
```
Comment (softwarecrash) [inline daly-bms-uart.h]: Sure its a good Idea? The Most of arduino librays usw This Name / function to Fall a global update

File: daly-bms-uart.h [hunk 3]
```diff
@@ -77,6 +75,15 @@ class Daly_BMS_UART
         // data from 0x97
         bool cellBalanceState[48];
         bool cellBalanceActive;
+        == == == =
```
Comment (softwarecrash) [inline daly-bms-uart.h:78]: Something ist wrong here

**Comment by softwarecrash** [comment]
sorry for the reup changes, something goes wrong with my keyboard

File: daly-bms-uart.h (hunk 3 above)
Comment (maland16) [reply daly-bms-uart.h]: That's my bad

File: daly-bms-uart.h (hunk 2 above)
Comment (maland16) [reply daly-bms-uart.h]: I agree with you

File: daly-bms-uart.h (hunk 3 above)
Comment (softwarecrash) [reply daly-bms-uart.h]: 😅 OK i was thinking was my Foul of May Keyboard, Sometimes it write curios letter

File: README.md [hunk 4]
```diff
@@ -48,9 +48,9 @@ Here's an overview of the commands that are supported by this library. See the f
 - Uncomment [this define](https://github.com/maland16/daly-bms-uart/blob/main/daly-bms-uart.h#L8) to get lots of debug printing to the Arduino Serial Monitor. I added these statements to help as I developed the code and ran into issues. Beyond that, I've done my best to comment extensively.  
 
 ## Future Improvements
+### Clean up/corrections  
+First & foremost there are a lot of wacky things in this repo that work, but are not done the best possible way. Cleaning things up and correcting bad coding practices would be huge.
```
Comment (softwarecrash) [inline README.md:52]: sorry about that, my coding skills are not top level, when you have the time, show me the right way and i will do the work

File: daly-bms-uart.cpp [hunk 5]
```diff
@@ -39,63 +38,451 @@ bool Daly_BMS_UART::Init()
...
+#endif
+        return false;
+    }
+    get.packVoltage = ((float)((this->my_rxBuffer[4] << 8) | this->my_rxBuffer[5]) / 10.0f);
+    get.packCurrent = ((float)(((this->my_rxBuffer[8] << 8) | this->my_rxBuffer[9]) - 30000) / 10.0f);
+    get.packSOC = ((float)((this->my_rxBuffer[10] << 8) | this->my_rxBuffer[11]) / 10.0f);
+#ifdef DALY_BMS_DEBUG
+
+    DEBUG_SERIAL.println("<DALY-BMS DEBUG> " + (String)get.packVoltage + "V, " + (String)get.packCurrent + "A, " + (String)get.packSOC + "SOC");
+    // debug current anomaly
```
Comment (softwarecrash) [inline daly-bms-uart.cpp]: following code will be removed when cleared what is with the bug that  randomly came a 2795A value

File: README.md [hunk 6]
```diff
@@ -48,9 +48,9 @@ Here's an overview of the commands that are supported by this library. See the f
...
-### Support for more commands
-There are a bunch of additional commands that are outlined in the protocol description document, so far I've only added support for the ones I intend to use.
 ### Redesign
-This library was purpose built for [this project](https://github.com/maland16/citicar-charger) and I made it a library in the hopes someone would get some utility out of it, but it could use to be redesigned from the ground up to be more capable/useful.
\ No newline at end of file
+This library was purpose built for [this project](https://github.com/maland16/citicar-charger) and I made it a library in the hopes someone would get some utility out of it, but it could use to be redesigned from the ground up to be more capable/useful. This library underwent a very significant redesign in PR#4 (shoutout to @softwarecrash) but I think it could still be a lot more user friendly.
```
Comment (softwarecrash) [inline README.md:56]: i have redsign it with the goal to easyer use in my opinion, whenever you have a idea to make it easyer to use let me know it

File: README.md (hunk 4 above)
Comment (maland16) [reply README.md:52]: There are wacky things done here by both of us haha, I didn't mean your changes specifically

File: daly-bms-uart.cpp [hunk 7]
```diff
@@ -39,63 +46,417 @@ bool Daly_BMS_UART::Init()
...
+
+    if (!this->receiveBytes())
+    {
+#ifdef DEBUG_SERIAL
+        DEBUG_SERIAL.println("<DALY-BMS DEBUG> Receive failed, Cell Balance State won't be modified!\n");
+#endif
+        return false;
+    }
+
+    for (size_t i = 0; i < 6; i++)
```
Comment (maland16) [inline daly-bms-uart.cpp:284]: @softwarecrash  Can you add a comment here explaining what's going on with these for loops, it's not immediately obvious

Comment (softwarecrash) [reply daly-bms-uart.cpp:284]: @maland16 sure,
the bms send for every cell a balance active bit, the inner for loop iterate a byte to bitwise and set the corespond bool flag on the array, the outer for iterate over the bytes in the bms message so it runs 6*8=48 the maximum numbers of cells. hope you can now comprehend it?

File: daly-bms-uart.h [hunk 8]
```diff
@@ -2,6 +2,10 @@
 #define DALY_BMS_UART_H
 
 #define XFER_BUFFER_LENGTH 13
+#define MIN_NUMBER_CELLS 1
+#define MAX_NUMBER_CELLS 48
```
Comment (softwarecrash) [inline daly-bms-uart.h:6]: The BMS send the value of max cell and temp Sensors, so i have used This Data to give a max Range, when This Data Not avaible from the BMS the following requests are useless, please Check it, i think variables are better than fixed values for this

**Pull request merged**